bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java

```diff
@@ -186,29 +186,36 @@ public void stopTestRun() {
 		int						port;
 		DataOutputStream		outStr;
 
-		public ControlThread(ILaunch launch, IJavaProject project, int port) {
+		public ControlThread(ILaunch launch, IJavaProject project, int port) throws IOException {
 			this.launch = launch;
 			this.project = project;
-			this.port = port;
+			listener = new ServerSocket(port);
```

Comment:
You also need to delete the `port` variable since it is no longer used. Also make `listener`, `launch`, and `project` variables final. The other variables should be volatile since they are written by one thread (run method) and read by another thread (close method).

Reply:
Done, done and done. I made the second two changes in a separate commit as they fix a separate issue. I also put a lock around the attempt to close `testRunSession` to avoid a race condition that could cause an NPE.